modules/videoio/src/cap.cpp

```diff
@@ -228,6 +235,128 @@ bool VideoCapture::open(const String& filename, int apiPreference, const std::ve
     return false;
 }
 
+bool VideoCapture::open(std::streambuf& stream, int apiPreference, const std::vector<int>& params)
+{
+    CV_INSTRUMENT_REGION();
+
+    if (isOpened())
+    {
+        release();
+    }
+
+    const VideoCaptureParameters parameters(params);
+    const std::vector<VideoBackendInfo> backends = cv::videoio_registry::getAvailableBackends_CaptureByBuffer();
+    for (size_t i = 0; i < backends.size(); i++)
+    {
+        const VideoBackendInfo& info = backends[i];
+        if (apiPreference == CAP_ANY || apiPreference == info.id)
+        {
+            if (!info.backendFactory)
+            {
+                CV_LOG_DEBUG(NULL, "VIDEOIO(" << info.name << "): factory is not available (plugins require filesystem support)");
+                continue;
+            }
+            CV_CAPTURE_LOG_DEBUG(NULL,
+                                 cv::format("VIDEOIO(%s): trying capture buffer ...",
+                                            info.name));
+            CV_Assert(!info.backendFactory.empty());
+            const Ptr<IBackend> backend = info.backendFactory->getBackend();
+            if (!backend.empty())
+            {
+                try
+                {
+                    icap = backend->createCapture(StreambufReadStream::create(stream), parameters);
```

Comment:
Who performs `seek(0, 0)` between trials?

Perhaps we should not allow `CAP_ANY` in this API.

Reply:
Good point, there is no explicit seek. Added a check that `apiPref != CAP_ANY`